File: services/pdf_service.py
from PyPDF2 import PdfReader, PdfWriter, PdfMerger
from fastapi import UploadFile
import os
import tempfile
import subprocess

from utils.file_utils import (
    save_uploaded_file,
    generate_split_filename,
    get_split_file_path,
    delete_file, get_merge_file_path
)
GS_PATH = r"C:\Program Files\gs\gs10.06.0\bin\gswin64c.exe"

QUALITY_MAP = {
    "high": "/prepress",
    "medium": "/printer",
    "low": "/ebook",
}

async def split_pdf_service(
    pdf: UploadFile,
    start_page: int,
    end_page: int
):
    input_path = save_uploaded_file(pdf)

    try:
        reader = PdfReader(input_path)
        total_pages = len(reader.pages)

        if start_page < 1 or end_page > total_pages or start_page > end_page:
            raise ValueError(
                f"Invalid page range! Total pages: {total_pages}"
            )

        writer = PdfWriter()
        for i in range(start_page - 1, end_page):
            writer.add_page(reader.pages[i])

        filename = generate_split_filename(start_page, end_page)
        output_path = get_split_file_path(filename)

        with open(output_path, "wb") as out:
            writer.write(out)

        return output_path, filename

    finally:
        # only delete uploaded temp file
        delete_file(input_path)

async def merge_pdf_service(pdf1: UploadFile, pdf2: UploadFile):
    path1 = save_uploaded_file(pdf1)
    path2 = save_uploaded_file(pdf2)

    merger = PdfMerger()

    try:
        reader1 = PdfReader(path1)
        reader2 = PdfReader(path2)

        if reader1.is_encrypted or reader2.is_encrypted:
            raise ValueError("One or both PDFs are password protected")

        merger.append(path1)
        merger.append(path2)

        filename = "merged.pdf"
        output_path = get_merge_file_path(filename)

        with open(output_path, "wb") as f:
            merger.write(f)

        return output_path, filename

    finally:
        merger.close()
        delete_file(path1)
        delete_file(path2)


async def compress_pdf_service(
    pdf: UploadFile,
    mode: str,
    value: int,
    quality: str,
):
    input_path = save_uploaded_file(pdf)

    try:
        output_name = f"compressed_{pdf.filename}"
        output_path = get_merge_file_path(output_name)

        gs_quality = QUALITY_MAP.get(quality, "/printer")

        cmd = [
            GS_PATH,
            "-sDEVICE=pdfwrite",
            "-dCompatibilityLevel=1.4",
            f"-dPDFSETTINGS={gs_quality}",
            "-dNOPAUSE",
            "-dQUIET",
            "-dBATCH",
            f"-sOutputFile={output_path}",
            input_path,
        ]

        subprocess.run(cmd, check=True)

        return output_path, output_name

    finally:
        delete_file(input_path)

File: src/modules/pdf/pdf.service.ts
import fs from "fs";
import path from "path";
import os from "os";
import { PDFDocument } from "pdf-lib";
import ghostscript from "ghostscript-node";
import { v4 as uuid } from "uuid";
import { execa } from "execa";

const SPLIT_DIR = "static/pdfs/split";
const MERGE_DIR = "static/pdfs/merge";

fs.mkdirSync(SPLIT_DIR, { recursive: true });
fs.mkdirSync(MERGE_DIR, { recursive: true });
const GS_COMMAND =
  os.platform() === "win32" ? "gswin64c" : "gs";

export const getPdfMetadata = async (pdfPath: string) => {
  const bytes = fs.readFileSync(pdfPath);
  const pdf = await PDFDocument.load(bytes);
  return pdf.getPageCount();
};

export const splitPdfService = async (
  pdfPath: string,
  start: number,
  end: number
) => {
  const srcPdf = await PDFDocument.load(fs.readFileSync(pdfPath));
  const total = srcPdf.getPageCount();

  if (start < 1 || end > total || start > end) {
    throw new Error(`Invalid page range! Total pages: ${total}`);
  }

  const newPdf = await PDFDocument.create();
  for (let i = start - 1; i < end; i++) {
    const [page] = await newPdf.copyPages(srcPdf, [i]);
    newPdf.addPage(page);
  }

  const filename = `${start}_${end}.pdf`;
  const outputPath = path.join(SPLIT_DIR, filename);

  fs.writeFileSync(outputPath, await newPdf.save());
  return { outputPath, filename };
};

export const mergePdfService = async (
  pdf1: string,
  pdf2: string
) => {
  const pdfA = await PDFDocument.load(fs.readFileSync(pdf1));
  const pdfB = await PDFDocument.load(fs.readFileSync(pdf2));

  const merged = await PDFDocument.create();
  for (const p of [pdfA, pdfB]) {
    const pages = await merged.copyPages(p, p.getPageIndices());
    pages.forEach(page => merged.addPage(page));
  }

  const filename = "merged.pdf";
  const outputPath = path.join(MERGE_DIR, filename);
  fs.writeFileSync(outputPath, await merged.save());

  return { outputPath, filename };
};

export const compressPdfService = async (
  input: string,
  quality: "high" | "medium" | "low"
) => {
  const outputPath = path.join(
    "static/pdfs/merge",
    `compressed_${uuid()}.pdf`
  );

  const QUALITY_MAP: Record<string, string> = {
    high: "/prepress",
    medium: "/printer",
    low: "/ebook"
  };

 await execa(GS_COMMAND, [
    "-sDEVICE=pdfwrite",
    "-dCompatibilityLevel=1.4",
    "-dPDFSETTINGS=/printer",
    "-dNOPAUSE",
    "-dQUIET",
    "-dBATCH",
    `-sOutputFile=${outputPath}`,
    input,
  ]);

  return outputPath;
};

